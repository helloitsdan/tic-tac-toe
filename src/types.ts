export enum BoardPosition {
  TOP_LEFT,
  TOP_CENTRE,
  TOP_RIGHT,
  MIDDLE_LEFT,
  MIDDLE_CENTRE,
  MIDDLE_RIGHT,
  BOTTOM_LEFT,
  BOTTOM_CENTRE,
  BOTTOM_RIGHT,
}

export enum PlayerPiece {
  PLAYER_ONE = 'PLAYER_ONE',
  PLAYER_TWO = 'PLAYER_TWO',
}

export enum MatchName {
  TOP_ROW = 'TOP_ROW',
  MIDDLE_ROW = 'MIDDLE_ROW',
  BOTTOM_ROW = 'BOTTOM_ROW',
  LEFT_COLUMN = 'LEFT_COLUMN',
  CENTRE_COLUMN = 'CENTRE_COLUMN',
  RIGHT_COLUMN = 'RIGHT_COLUMN',
  DIAGONAL_LEFT_TO_RIGHT = 'DIAGONAL_LEFT_TO_RIGHT',
  DIAGONAL_RIGHT_TO_LEFT = 'DIAGONAL_RIGHT_TO_LEFT',
}

export interface PlayerType {
  piece: PlayerPiece
  name: string
}

export type MatchTuple = [MatchName, BoardPosition[]]

export interface PlayerMatchType {
  match: MatchTuple
  piece: PlayerPiece
}

export type GameMovesType = Record<PlayerPiece, BoardPosition[]>

export interface GameStateType {
  players: PlayerType[]
  moves: GameMovesType
  winner: PlayerMatchType
}
