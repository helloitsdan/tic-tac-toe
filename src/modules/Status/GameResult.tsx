import React, { FunctionComponent, useMemo } from 'react'
import { useSelector } from 'react-redux'
import { getWinner, getPlayers } from '../../store/selectors'

const GameResult: FunctionComponent = () => {
  const winner = useSelector(getWinner)
  const players = useSelector(getPlayers)

  const player = useMemo(() => {
    if (!winner) {
      return null
    }

    return players.all.find(p => p.piece === winner!.player)
  }, [winner, players])

  return player ? (
    <div data-testid="GameResult/Player">
      <span data-testid="GameResult/PlayerName">{player.name}</span> wins!
    </div>
  ) : (
    <div data-testid="GameResult/Draw">A draw...</div>
  )
}

export default GameResult
