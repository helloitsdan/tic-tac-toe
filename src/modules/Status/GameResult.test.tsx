import React from 'react'
import { render } from '@testing-library/react'

import GameResult from './GameResult'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { PlayerId } from '../../types'

describe('GameResult', () => {
  const testPlayers = {
    active: PlayerId.PLAYER_ONE,
    all: [
      { piece: PlayerId.PLAYER_ONE, name: 'Test Player One' },
      { piece: PlayerId.PLAYER_TWO, name: 'Test Player Two' },
    ]
  }

  it('renders', () => {
    const testStore = createStore(() => ({
      winner: {
        player: PlayerId.PLAYER_ONE,
        match: [],
      },
      players: testPlayers,
    }))

    const { asFragment } = render(
      <Provider store={testStore}>
        <GameResult />
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders a message for a draw', () => {
    const testStore = createStore(() => ({
      winner: null,
      players: testPlayers,
    }))

    const { getByTestId } = render(
      <Provider store={testStore}>
        <GameResult />
      </Provider>
    )
    
    expect(getByTestId('GameResult/Draw')).not.toBeNull()
  })

  it('renders a message when player one wins', () => {
    const testStore = createStore(() => ({
      winner: {
        player: PlayerId.PLAYER_ONE,
        match: [],
      },
      players: testPlayers,
    }))

    const { getByText } = render(
      <Provider store={testStore}>
        <GameResult />
      </Provider>
    )
    
    expect(getByText(testPlayers.all[0].name)).not.toBeNull()
  })

  it('renders a message when player two wins', () => {
    const testStore = createStore(() => ({
      winner: {
        player: PlayerId.PLAYER_TWO,
        match: [],
      },
      players: testPlayers,
    }))

    const { getByText } = render(
      <Provider store={testStore}>
        <GameResult />
      </Provider>
    )
    
    expect(getByText(testPlayers.all[1].name)).not.toBeNull()
  })
})
