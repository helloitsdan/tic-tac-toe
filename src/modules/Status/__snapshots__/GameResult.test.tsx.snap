// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameResult renders 1`] = `
<DocumentFragment>
  <div
    data-testid="GameResult/Player"
  >
    <span
      data-testid="GameResult/PlayerName"
    >
      Test Player One
    </span>
     wins!
  </div>
</DocumentFragment>
`;
